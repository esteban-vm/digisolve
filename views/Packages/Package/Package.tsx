import type { Component, PropsWithLink } from '@/types'
import { Button, Icon } from '@/components'
import { formatPrice } from '@/helpers'
import { styled, mediaQuery } from '@/styles'

export type PackageProps = PropsWithLink<{
  title: string
  price: number
  duration: number
  description: string
  features: [string, string, string, string, string]
}>

const PackageComponent: Component<PackageProps> = ({
  id,
  title,
  price,
  duration,
  description,
  features,
  link,
  ...rest
}) => {
  const formattedPrice = formatPrice(price)

  return (
    <article aria-labelledby={id} {...rest}>
      <div className='section'>
        <h3 id={id}>{title}</h3>
        <p>
          {formattedPrice} <span>{duration} months</span>
        </p>
        <p>{description}</p>
      </div>
      <div className='section'>
        <ul>
          {features.map((feature, index) => (
            <li key={index}>
              <Icon name='check' size={1.2} className='icon' />
              {feature}
            </li>
          ))}
        </ul>
      </div>
      <div className='section'>
        <Button text='Sign up now' link={link} className='btn' isFull />
      </div>
    </article>
  )
}

const Package = styled(PackageComponent)`
  background-color: var(--color-white);
  width: 90%;
  margin: 0 auto 30px;
  box-shadow: 0 3px 2px var(--color-shadow);
  text-align: left;

  ${mediaQuery('md')} {
    width: 100%;
  }

  ${mediaQuery('xs')} {
    width: 80%;
  }

  h3 {
    font-size: 150%;
    color: var(--color-grey);
    font-weight: 400;
    margin-bottom: 50px;
    text-transform: capitalize;

    ${mediaQuery('md')} {
      margin-bottom: 25px;
    }
  }

  p {
    :first-of-type {
      color: var(--color-primary);
      font-size: 300%;
      font-weight: 300;
      margin-bottom: 10px;

      span {
        font-size: 30%;
        font-weight: 500;
      }
    }

    :last-of-type {
      color: var(--color-grey);
      font-size: 80%;
      margin-bottom: 10px;
    }
  }

  ul {
    li {
      color: var(--color-grey);
      padding: 5px 0;
    }
  }

  .section {
    padding: 15px;
    border-bottom: 1px solid var(--color-shadow);

    :first-of-type {
      background-color: var(--color-lighter);
    }

    :last-of-type {
      border: 0;
      text-align: center;
    }
  }

  .icon {
    display: inline-block;
    margin-right: 10px;
  }

  .btn {
    margin: 15px 0;
  }
`

export default Package
