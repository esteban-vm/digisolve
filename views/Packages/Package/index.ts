export { default, type PackageProps } from './Package'
