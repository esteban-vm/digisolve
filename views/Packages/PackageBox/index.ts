export { default, type PackageProps } from './PackageBox'
