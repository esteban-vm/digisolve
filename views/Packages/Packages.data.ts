import type { PackageProps } from './Package'

const description = 'Ratione similique aut incidunt repellendus excepturi ut commodi laboriosam.'

const packages: PackageProps[] = [
  {
    id: 'd7e645e0-e825-11ed-b24d-893619994521',
    title: 'starter',
    price: 699,
    duration: 3,
    description,
    features: [
      'SEO, SMO and PPC',
      '2 Blogs per month',
      'Facebook & Twitter Marketing',
      'Monthly Performance',
      'Traffic Report',
    ],
  },
  {
    id: 'df095c40-e825-11ed-b24d-893619994521',
    title: 'corporate',
    price: 2999,
    duration: 6,
    description,
    features: [
      'SEO, SMO and PPC',
      '4 Blogs per month',
      'All Social Media Channels',
      'Monthly Performance',
      'Traffic Report',
    ],
  },
  {
    id: 'e21563c0-e825-11ed-b24d-893619994521',
    title: 'enterprice',
    price: 4999,
    duration: 6,
    description,
    features: [
      'Personalized Marketing',
      '8 Blogs per month',
      'All Social Media Channels',
      'Monthly Performance',
      'Traffic Report',
    ],
  },
]

export default packages
