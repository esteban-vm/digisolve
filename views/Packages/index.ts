export { default as Packages } from './Packages'
