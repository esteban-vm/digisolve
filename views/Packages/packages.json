[
  {
    "title": "starter",
    "price": 699,
    "duration": 3,
    "features": [
      "SEO, SMO and PPC",
      "2 Blogs per month",
      "Facebook & Twitter Marketing",
      "Monthly Performance",
      "Traffic Report"
    ]
  },
  {
    "title": "corporate",
    "price": 2999,
    "duration": 6,
    "features": [
      "SEO, SMO and PPC",
      "4 Blogs per month",
      "All Social Media Channels",
      "Monthly Performance",
      "Traffic Report"
    ]
  },
  {
    "title": "enterprice",
    "price": 4999,
    "duration": 6,
    "features": [
      "Personalized Marketing",
      "8 Blogs per month",
      "All Social Media Channels",
      "Monthly Performance",
      "Traffic Report"
    ]
  }
]
