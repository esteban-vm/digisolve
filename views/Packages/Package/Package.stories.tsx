import type { Meta, StoryObj } from '@/stories'
import { Grid } from '@/components'
import { styled } from '@/styles'
import { packages } from '../Packages'
import Package from './Package'

const Wrapper = styled.div`
  width: 75vw;
`

export default {
  argTypes: {
    benefits: {
      control: false,
    },
  },
  component: Package,
  title: 'Views/Packages/Packages',
  decorators: [
    (Story, context) => {
      return (
        <Wrapper>
          <Grid.Row>
            <Grid.Col isOneThird>&nbsp;</Grid.Col>
            <Grid.Col isOneThird>
              <Story {...context} />
            </Grid.Col>
            <Grid.Col isOneThird>&nbsp;</Grid.Col>
          </Grid.Row>
        </Wrapper>
      )
    },
  ],
  parameters: {
    layout: 'centered',
  },
} as Meta<typeof Package>

const Examples: StoryObj<typeof Package>[] = []
for (const pack of packages) Examples.push({ args: pack })

export const Example1 = Examples[0]
export const Example2 = Examples[1]
export const Example3 = Examples[2]
