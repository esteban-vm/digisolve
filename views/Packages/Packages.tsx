import type { Component } from '@/types'
import { useId } from 'react'
import { Grid } from '@/components'
import { styled, mediaQuery } from '@/styles'
import packages from './Packages.data'
import Package from './Package'

const PackagesComponent: Component = (props) => {
  const nameID = useId()

  return (
    <section id='packages' aria-labelledby={nameID} {...props}>
      <Grid.Row>
        <h2 id={nameID}>Check out our most popular packages</h2>
      </Grid.Row>
      <Grid.Row>
        {packages.map((_package) => (
          <Grid.Col key={_package.id} isOneThird>
            <Package {..._package} />
          </Grid.Col>
        ))}
      </Grid.Row>
    </section>
  )
}

const Packages = styled(PackagesComponent)`
  text-align: center;
  background-color: var(--color-light);

  h2 {
    color: var(--color-black);
    font-size: 250%;
    margin-bottom: 20px;
    margin-top: 150px;
    text-transform: capitalize;

    ${mediaQuery('lg')} {
      margin-top: 90px;
    }

    ${mediaQuery('sm')} {
      margin-top: 60px;
    }
  }
`

export default Packages
