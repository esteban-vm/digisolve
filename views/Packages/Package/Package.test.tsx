import { render, cleanup, screen, create } from '@/tests'
import { formatPrice } from '@/helpers'
import { packages } from '../Packages'
import Package from './Package'

describe('ðŸ§ª PACKAGE:', () => {
  const [testPackage] = packages

  describe('display tests:', () => {
    let parent: HTMLElement

    beforeEach(() => {
      render(<Package {...testPackage} />)
      parent = screen.getByRole('article')
    })

    afterEach(cleanup)

    it('should display the heading', () => {
      const heading = screen.getByRole('heading', { name: testPackage.heading, level: 4 })
      expect(heading).toBeInTheDocument()
      expect(heading).toBeVisible()
      expect(parent).toContainElement(heading)
    })

    it('should display the price', () => {
      const price = screen.getByText(formatPrice(testPackage.price))
      expect(price).toBeInTheDocument()
      expect(price).toBeVisible()
      expect(parent).toContainElement(price)
    })

    it('should display the duration', () => {
      const duration = screen.getByText(`${testPackage.duration} months`)
      expect(duration).toBeInTheDocument()
      expect(duration).toBeVisible()
      expect(parent).toContainElement(duration)
    })

    it('should display the description', () => {
      const description = screen.getByText(/^lorem ipsum/i)
      expect(description).toBeInTheDocument()
      expect(description).toBeVisible()
      expect(parent).toContainElement(description)
    })

    it('should display the benefits', () => {
      const list = screen.getByRole('list')
      const benefits = screen.getAllByRole('listitem')
      const icons = screen.getAllByRole('img', { hidden: true })
      expect(icons.length).toBe(benefits.length)

      for (let index = 0; index < benefits.length; index++) {
        const benefit = benefits[index]
        expect(benefit).toBeInTheDocument()
        expect(benefit).toBeVisible()
        expect(benefit).toContainElement(icons[index])
        expect(benefit).toHaveTextContent(testPackage.benefits[index])
        expect(list).toContainElement(benefit)
      }

      expect(parent).toContainElement(list)
    })

    it('should display the button', () => {
      const button = screen.getByRole('button', { name: /^sign up now$/i })
      expect(button).toBeInTheDocument()
      expect(button).toBeVisible()
      expect(parent).toContainElement(button)
    })
  })

  describe('style tests', () => {
    it('should render with correct styles', () => {
      const tree = create(<Package {...testPackage} />)
      expect(tree).toMatchSnapshot()
    })
  })
})
