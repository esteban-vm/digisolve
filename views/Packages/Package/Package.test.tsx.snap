// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ðŸ§ª PACKAGE: style tests: should render with correct styles 1`] = `
.emotion-0 {
  background-color: var(--color-white);
  width: 90%;
  margin: 0 auto 30px;
  box-shadow: 0 3px 2px var(--color-shadow);
  text-align: left;
}

@media only screen and (max-width: 1024px) {
  .emotion-0 {
    width: 100%;
  }
}

@media only screen and (max-width: 480px) {
  .emotion-0 {
    width: 80%;
  }
}

.emotion-0 h3 {
  font-size: 150%;
  color: var(--color-grey);
  font-weight: 400;
  margin-bottom: 50px;
  text-transform: capitalize;
}

@media only screen and (max-width: 1024px) {
  .emotion-0 h3 {
    margin-bottom: 25px;
  }
}

.emotion-0 p:first-of-type {
  color: var(--color-primary);
  font-size: 300%;
  font-weight: 300;
  margin-bottom: 10px;
}

.emotion-0 p:first-of-type span {
  font-size: 30%;
  font-weight: 500;
}

.emotion-0 p:last-of-type {
  color: var(--color-grey);
  font-size: 80%;
  margin-bottom: 10px;
}

.emotion-0 ul li {
  color: var(--color-grey);
  padding: 5px 0;
}

.emotion-0 .section {
  padding: 15px;
  border-bottom: 1px solid var(--color-shadow);
}

.emotion-0 .section:first-of-type {
  background-color: var(--color-lighter);
}

.emotion-0 .section:last-of-type {
  border: 0;
  text-align: center;
}

.emotion-0 .icon {
  display: inline-block;
  margin-right: 10px;
}

.emotion-0 .btn {
  margin: 15px 0;
}

.emotion-1 {
  cursor: pointer;
  display: inline-block;
  -webkit-text-decoration: none;
  text-decoration: none;
  text-transform: capitalize;
  padding: 10px 30px;
  border-radius: 200px;
  font-weight: 500;
  outline: none;
  -webkit-transition: background-color 500ms;
  transition: background-color 500ms;
  border: 2px solid var(--color-primary);
  background-color: var(--color-primary);
  color: var(--color-white);
}

.emotion-1:hover {
  background-color: var(--color-dark-primary);
}

<article
  aria-labelledby="d7e645e0-e825-11ed-b24d-893619994521"
  className="emotion-0"
>
  <div
    className="section"
  >
    <h3
      id="d7e645e0-e825-11ed-b24d-893619994521"
    >
      starter
    </h3>
    <p>
      $699
       
      <span>
        3
         months
      </span>
    </p>
    <p>
      Ratione similique aut incidunt repellendus excepturi ut commodi laboriosam.
    </p>
  </div>
  <div
    className="section"
  >
    <ul>
      <li>
        <svg
          aria-hidden={true}
          className="icon"
          color="var(--color-primary)"
          fill="currentColor"
          height="1.2rem"
          role="img"
          viewBox="0 0 512 512"
          width="1.2rem"
        >
          <path
            d="M470.6 105.4c12.5 12.5 12.5 32.8 0 45.3l-256 256c-12.5 12.5-32.8 12.5-45.3 0l-128-128c-12.5-12.5-12.5-32.8 0-45.3s32.8-12.5 45.3 0L192 338.7l233.4-233.3c12.5-12.5 32.8-12.5 45.3 0z"
          />
        </svg>
        SEO, SMO and PPC
      </li>
      <li>
        <svg
          aria-hidden={true}
          className="icon"
          color="var(--color-primary)"
          fill="currentColor"
          height="1.2rem"
          role="img"
          viewBox="0 0 512 512"
          width="1.2rem"
        >
          <path
            d="M470.6 105.4c12.5 12.5 12.5 32.8 0 45.3l-256 256c-12.5 12.5-32.8 12.5-45.3 0l-128-128c-12.5-12.5-12.5-32.8 0-45.3s32.8-12.5 45.3 0L192 338.7l233.4-233.3c12.5-12.5 32.8-12.5 45.3 0z"
          />
        </svg>
        2 Blogs per month
      </li>
      <li>
        <svg
          aria-hidden={true}
          className="icon"
          color="var(--color-primary)"
          fill="currentColor"
          height="1.2rem"
          role="img"
          viewBox="0 0 512 512"
          width="1.2rem"
        >
          <path
            d="M470.6 105.4c12.5 12.5 12.5 32.8 0 45.3l-256 256c-12.5 12.5-32.8 12.5-45.3 0l-128-128c-12.5-12.5-12.5-32.8 0-45.3s32.8-12.5 45.3 0L192 338.7l233.4-233.3c12.5-12.5 32.8-12.5 45.3 0z"
          />
        </svg>
        Facebook & Twitter Marketing
      </li>
      <li>
        <svg
          aria-hidden={true}
          className="icon"
          color="var(--color-primary)"
          fill="currentColor"
          height="1.2rem"
          role="img"
          viewBox="0 0 512 512"
          width="1.2rem"
        >
          <path
            d="M470.6 105.4c12.5 12.5 12.5 32.8 0 45.3l-256 256c-12.5 12.5-32.8 12.5-45.3 0l-128-128c-12.5-12.5-12.5-32.8 0-45.3s32.8-12.5 45.3 0L192 338.7l233.4-233.3c12.5-12.5 32.8-12.5 45.3 0z"
          />
        </svg>
        Monthly Performance
      </li>
      <li>
        <svg
          aria-hidden={true}
          className="icon"
          color="var(--color-primary)"
          fill="currentColor"
          height="1.2rem"
          role="img"
          viewBox="0 0 512 512"
          width="1.2rem"
        >
          <path
            d="M470.6 105.4c12.5 12.5 12.5 32.8 0 45.3l-256 256c-12.5 12.5-32.8 12.5-45.3 0l-128-128c-12.5-12.5-12.5-32.8 0-45.3s32.8-12.5 45.3 0L192 338.7l233.4-233.3c12.5-12.5 32.8-12.5 45.3 0z"
          />
        </svg>
        Traffic Report
      </li>
    </ul>
  </div>
  <div
    className="section"
  >
    <button
      className="btn emotion-1"
      onClick={[Function]}
      type="button"
    >
      Sign up now
    </button>
  </div>
</article>
`;
